src/drivers/efi/capsules.c

```diff
@@ -0,0 +1,742 @@
+{
+	struct memory_range buffer = {0};
+
+	const struct range_entry *r;
+	memranges_each_entry(r, &memory_map) {
+		if (range_entry_tag(r) != BM_MEM_RAM)
+			continue;
+
+		resource_t base = range_entry_base(r);
+		resource_t size = range_entry_size(r);
+
+		if (base > 4ULL * GiB)
+			break;
+
+		/* Don't allocate the buffer too low. */
+		if (base < 1 * MiB) {
+			if (size <= 1 * MiB)
+				continue;
+
+			base += 1 * MiB;
+			size -= 1 * MiB;
+		}
+
+		if (size < total_data_size)
+			continue;
+		if (base + total_data_size > 4ULL * GiB)
+			continue;
+
+		buffer.base = base;
+		buffer.len = total_data_size;
+
+		/* Keep looking for the highest block to not create troubles for payloads. */
+	}
+
+	return buffer;
+}
+
+/* Puts capsules into continuous physical memory. */
+static void coalesce_capsules(struct block_descr block_chain, uint8_t *target)
+{
+	struct block_descr block = block_chain;
+	uint8_t *capsule_start = NULL;
+	uint32_t size_left = 0;
+
+	/* No safety checks in this function, as all of them were done earlier. */
+
+	load_block(&block);
+	for (; !is_final_block(&block); advance_block(&block), load_block(&block)) {
+		/* Advance over a continuation. */
+		if (!is_data_block(&block))
+			continue;
+
+		/* This must be the first block of a capsule. */
+		if (size_left == 0) {
+			const EFI_CAPSULE_HEADER *capsule_hdr =
+				map_range(block.addr, sizeof(*capsule_hdr));
+			size_left = capsule_hdr->CapsuleImageSize;
+			capsule_start = target;
+		}
+
+		uint64_t addr = block.addr;
+		uint64_t data_left = block.len;
+		while (data_left != 0) {
+			uint64_t piece_len = MIN(data_left, 2 * MiB - (addr % 2 * MiB));
+			void *data = map_range(addr, piece_len);
+
+			memcpy(target, data, piece_len);
+
+			target += piece_len;
+			addr += piece_len;
+			data_left -= piece_len;
+		}
+
+		size_left -= block.len;
+
+		/* This must be the last block of a capsule, record it. */
+		if (size_left == 0) {
+			/* If we can just ignore corrupted capsules, then we can simply
+			   drop those which don't fit. */
+			if (uefi_capsule_count == MAX_CAPSULES) {
+				printk(BIOS_WARNING,
+				       "capsules: ignoring all capsules after #%d.\n",
+				       MAX_CAPSULES);
+				break;
+			}
+
+			uefi_capsules[uefi_capsule_count].base = (uintptr_t)capsule_start;
+			uefi_capsules[uefi_capsule_count].len = block.len;
+			uefi_capsule_count++;
+
+			/* This is to align start of the next capsule (assumes that
+			   initial value of target was suitably aligned). */
+			if (!IS_ALIGNED(block.len, CAPSULE_ALIGNMENT))
+				target += ALIGN_UP(block.len, CAPSULE_ALIGNMENT) - block.len;
+		}
+	}
+
+	printk(BIOS_INFO, "capsules: found %d capsule(s).\n", uefi_capsule_count);
+}
+
+void efi_parse_capsules(void)
+{
+	/* EDK2 starts with 20 items and then grows the list, but it's unlikely
+	   to be necessary in practice. */
+	enum { MAX_CAPSULE_BLOCKS = MAX_CAPSULES };
+
+	struct region_device rdev;
+	if (!smmstore_lookup_region(&rdev)) {
```

Comment:
Something is wrong around here:

```
FMAP: area SMMSTORE found @ 0 (262144 bytes)
capsules: no SMMSTORE region, no update capsules.
```

Reply:
Inverted condition.  Must have messed it up while addressing review.  There was another issue as well.

https://github.com/Dasharo/coreboot/compare/11de72c3ff2f755a0fb7ac45d38f93c5bf7a68c9..628cce5b62a2df64b6150e4b63ef1fb715b1c6ea